### PR TITLE
[GSoC] Add block quantized models 

### PR DESCRIPTION
**Pull request by DaniAffCH**
This PR introduces block-quantized versions for most of the opencv_zoo models.
All the models have been quantized using a block size of 64, as this configuration demonstrated good performance empirically.

Additionally, the block quantization tool has been enhanced to handle more cases:

- The tool now supports weights categorized as `constant` in ONNX (previously, it only supported `initializers`).
- It now properly handles blocks where all elements are identical.
- Data type saturation has been added to prevent overflow during quantization.
- The error metric has been updated to be independent of the weights' order of magnitude
- A new verbose mode has been added to assist in troubleshooting.

Finally, the benchmark tool has been modified to support block quantized models.

The following table contains block quantization statistics of the quantized models

| Model    | Original Size (KB) | Block Quantized Size (KB) |
| -------- | ------- | ------- |
| face_detection_yunet | 227.14 | 119.62 | 
| face_recognition_sface | 37,789.41 | 10,417.82 | 
| facial_expression_recognition_mobilefacenet | 4,679.58 | 1,344.44 | 
| handpose_estimation_mediapipe | 4,003.54 | 1,199.43 | 
| human_segmentation_pphumanseg | 6,019.47 | 1,694.07 | 
| image_classification_mobilenetv1 | 16,494.27 | 4,491.59 | 
| image_classification_mobilenetv2 | 13,637.28 | 3,782.18 | 
| image_classification_ppresnet50 | 100,163.12 | 27,435.20 | 
| license_plate_detection_lpd_yunet | 4,049.04 | 1,158.07 | 
| object_detection_nanodet | 3,711.87 | 1,097.62 | 
| object_detection_yolox | 35,017.58 | 9,516.03 | 
| object_tracking_vittrack | 697.97 | 264.97 | 
| optical_flow_estimation_raft | 62,616.54 | 47,700.30 | 
| palm_detection_mediapipe | 3,814.19 | 1,141.94 | 
| person_detection_mediapipe | 11,709.14 | 3,400.44 | 
| person_reid_youtu | 104,373.44 | 28,518.79 | 
| pose_estimation_mediapipe | 5,426.99 | 1,655.17 | 
| text_detection_en_ppocrv3 | 2,366.69 | 835.33 | 
| text_detection_cn_ppocrv3 | 2,366.69 | 835.33 | 
| text_recognition_CRNN_CN | 71,100.74 | 28,346.08 | 
| text_recognition_CRNN_CH | 63,385.71 | 26,257.37 | 

The tables below summarize the metrics change between the original fp32 model, block quantized and int8 quantized version:

| Models      | Accuracy |
| ----------- | -------- |
| SFace       | 0.9940   |
| **SFace block** | **0.9942**   |
| SFace quant | 0.9932   | 

| Models      | Easy AP | Medium AP | Hard AP |
| ----------- | ------- | --------- | ------- |
| YuNet       | 0.8844  | 0.8656    | 0.7503  |
| **YuNet block** | **0.8845**  | **0.8652**    | **0.7504**  |
| YuNet quant | 0.8810  | 0.8629    | 0.7503  |

| Models             | Accuracy       | mIoU          |
| ------------------ | -------------- | ------------- |
| PPHumanSeg         | 0.9656         | 0.9164        |
| **PPHumanSeg block**   | **0.9655**         | **0.9162**        |
| PPHumanSeg quant   | 0.7285         | 0.3642        |

| Models             | Top-1 Accuracy | Top-5 Accuracy |
| ------------------ | -------------- | -------------- |
| MobileNet V1       | 67.64          | 87.97          |
| **MobileNet V1 block** | **67.21**          | **87.62**          |
| MobileNet V1 quant | 55.53          | 78.74          |
| MobileNet V2       | 69.44          | 89.23          |
| **MobileNet V2 block** | **68.66**          | **88.90**          |
| MobileNet V2 quant | 68.37          | 88.56          |
 
| Models          | Top-1 Accuracy | Top-5 Accuracy |
| --------------- | -------------- | -------------- |
| PP-ResNet       | 82.28          | 96.15          |
| **PP-ResNet block** | **82.27**          | **96.15**          |
| PP-ResNet quant | 0.22           | 0.96           |

The following models haven't been quantized:
- **image_segmentation_efficientsam**: Not compliant with the ONNX standard (#269).
- **text_recognition_CRNN_EN**: Even a minor error prevents the model from correctly predicting the characters, despite accurately predicting the text bounding box. The reason for this issue remains unclear, as the Chinese version of the model works properly.

### CODE REVIEW

**Comment by vpisarev** [comment]
@fengyuentau, when it's expected to be merged? I believe, the patch is very useful.

**Comment by fengyuentau** [comment]
> @fengyuentau, when it's expected to be merged? I believe, the patch is very useful.

We can, but since block quantization in opencv does not have any acceleration in terms of inference speed, merging this patch only increases the size of zoo and we are not gaining any benifit from it for now. Merging is postponed until block quantization works practically better in opencv.

**Comment by vpisarev** [comment]
I don't quite get it. We don't see any performance _degradation_ with block-wise quantized models _and_ the models get smaller, not bigger. That is, we get smaller size + same speed. But if you are still in doubt, let's discuss it next week at our meeting, pls, include it into the agenda

**Comment by fengyuentau** [comment]
> I don't quite get it. We don't see any performance _degradation_ with block-wise quantized models _and_ the models get smaller, not bigger. That is, we get smaller size + same speed. But if you are still in doubt, let's discuss it next week at our meeting, pls, include it into the agenda

Okay, lets do it in the next meeting.

**Comment by fengyuentau** [comment]
@DaniAffCH Thank you for all the effort on this pull request. We decided to merge this pull request with the following changes:

1. Could you report the accuracy of text detection and recoginition as well? Even though they may not have a good number, we still want to know about this since it is here in the pull request.
2. Change model naming suffix from `_blocked` to `_int8bq`. Add a note in each `README` explaining models with suffix `_int8bq` is block-quantized in int8 precision.

**Comment by fengyuentau** [comment]
Also need to provide each command that you use to generate the block-quantized models. Just want to ensure reproducibility.

**Comment by fengyuentau** [comment]
@DaniAffCH Do you plan to push commits to finalize this PR? If no, I will merge this one first then do it in the subsequent PRs.

**Comment by DaniAffCH** [comment]
Yes I'll definitely address your comments. I've been busy with other projects, but now I can finalize this PR.

**Comment by DaniAffCH** [comment]
I've just updated the file suffixes and the related README files. 

> Could you report the accuracy of text detection and recoginition as well? Even though they may not have a good number, we still want to know about this since it is here in the pull request.

Regarding text detection, I couldn't find a suitable evaluation script in [eval](https://github.com/opencv/opencv_zoo/tree/main/tools/eval), so I don't know how to test it.

Regarding text recognition, I decided not to include the English version because of a severe drop in accuracy. Such a drop in accuracy doesn't occur in the Chinese version, so I decided to include only `text_recognition_CRNN_CH` `text_recognition_CRNN_CN`.
However, the evaluation script only supports IIIT5K and ICDAR2003 datasets containing English words, so it's not possible to test the Chinese models on them.

File: models/__init__.py [hunk 1]
```diff
@@ -72,6 +75,7 @@ def register(self, item):
             fp32=fp32_model_paths,
             fp16=fp16_model_paths,
             int8=int8_model_paths,
+            bint8=bint8_model_paths
```
Comment (fengyuentau) [inline models/__init__.py:78]: int8bq=int8bq_model_paths

**Comment by DaniAffCH** [comment]
Done!

**Review by fengyuentau — APPROVED** [review]
Great! Thank you! 👍

**Pull request merged**